oak/server/rust/oak_runtime/src/time/mod.rs

```diff
@@ -0,0 +1,383 @@
+                None
+            }
+        }
+    }
+
+    /// Makes a Roughtime request to a server.
+    async fn get_roughtime_from_server(
+        &self,
+        server: &RoughtimeServerSpec,
+    ) -> Result<Interval, RoughtimeError> {
+        let nonce = create_nonce()?;
+        let request = make_request(&nonce)?;
+        let mut failed_attempts = 0;
+        let response = loop {
+            match self.send_roughtime_request(server, &request).await {
+                Err(RoughtimeError::TimeoutError) => {
+                    failed_attempts += 1;
+                    if failed_attempts >= self.server_retries {
+                        break Err(RoughtimeError::TimeoutError);
+                    }
+                }
+                result => break result,
+            }
+        }?;
+        let ParsedResponse {
+            verified,
+            midpoint,
+            radius,
+        } = ResponseHandler::new(
+            base64::decode(server.public_key_base64.as_bytes())?,
+            RtMessage::from_bytes(response.as_ref())?,
+            nonce.to_owned(),
+        )?
+        .extract_time()?;
+        debug!(
+            "Roughtime from {}: midpoint={}, radius={}, verified={}",
+            server.name, midpoint, radius, verified
+        );
+
+        if verified {
+            if radius <= self.max_radius_microseconds {
+                if midpoint > radius.into() {
+                    Ok(Interval {
+                        min: midpoint.saturating_sub(radius.into()),
+                        max: midpoint.saturating_add(radius.into()),
+                    })
+                } else {
+                    Err(RoughtimeError::MidPointTooSmall)
+                }
+            } else {
+                Err(RoughtimeError::RadiusTooLarge)
+            }
+        } else {
+            Err(RoughtimeError::InvalidSignature)
+        }
+    }
+
+    /// Sends a request to a Roughtime server using UDP.
+    async fn send_roughtime_request(
+        &self,
+        server: &RoughtimeServerSpec,
+        request: &[u8],
+    ) -> Result<Vec<u8>, RoughtimeError> {
+        let remote_addr = (&server.host[..], server.port)
+            .to_socket_addrs()?
+            .next()
+            .unwrap();
+        let local_address: SocketAddr = if remote_addr.is_ipv4() {
+            "0.0.0.0:0"
+        } else {
+            "[::]:0"
+        }
+        .parse()
+        .unwrap();
+        let mut socket = UdpSocket::bind(local_address).await?;
+        socket.connect(&remote_addr).await?;
+        match timeout(
+            Duration::from_secs(self.timeout_seconds),
+            socket.send(request),
+        )
+        .await
+        {
+            Err(error) => {
+                warn!("Timed out sending request to {}: {}", server.name, error);
+                return Err(RoughtimeError::TimeoutError);
+            }
+            Ok(result) => {
+                result?;
```

Comment:
What is result? Don't we need to do anything with it? Check its value or something?